site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+                }
+            }
+            if ($this->team_assignment) {
+                if ($team_lock_date === null) {
+                    $errors['team_lock_date'] = "Value must not be null!";
+                }
+            }
+        } else {
+            // The only check if its not an electronic gradeable
+            if (!($ta_view_start_date === null || $grade_released_date === null) && $ta_view_start_date > $grade_released_date) {
+                $errors['g_grade_released_date'] = 'Grades Released Date must be later than the TA Beta Testing Date';
+            }
+        }
+
+        if (count($errors) === 0)
+            return null;
+        return $errors;
+    }
+
+    public function setDates(array $dates)
+    {
+        $errors = $this->validateDates($dates);
+
+        if ($errors !== null) {
+            throw new AggregateException("Date validation failed!", $errors);
+        }
+
+        $this->ta_view_start_date = $dates['ta_view_start_date'];
+        $this->grade_start_date = $dates['grade_start_date'];
+        $this->grade_released_date = $dates['grade_released_date'];
+        $this->grade_locked_date = $dates['grade_locked_date'];
+
+        if($this->type === GradeableType::ELECTRONIC_FILE) {
+            if($this->team_assignment) {
+                $this->team_lock_date = $dates['team_lock_date'];
+            }
+            $this->submission_open_date = $dates['submission_open_date'];
+            $this->submission_due_date = $dates['submission_due_date'];
+            $this->late_days = $dates['late_days'];
+        }
+    }
+    private function setTaViewStartDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeStartDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeReleasedDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeLockedDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setTeamLockDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setSubmissionOpenDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setSubmissionDueDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+
+    private function setId($id)
+    {
+        preg_match('/^[a-zA-Z0-9_-]*$/', $id, $matches, PREG_OFFSET_CAPTURE);
+        if (count($matches) === 0) {
```

Comment:
If `Gradeable::setId()` is private, should the other classes also have private `setId()`?

Reply:
fixed